src/core/detail/btree_internal.h

```diff
@@ -0,0 +1,458 @@
+  // Returns retired BTreeNode* if children got merged and this parent node's children
+  // count decreased, otherwise, we return nullptr (rebalanced).
+  BTreeNode* MergeOrRebalanceChild(unsigned pos);
+
+  void Validate(Key_t upper_bound) const;
+
+ private:
+  void RebalanceChildToLeft(unsigned child_pos, unsigned count);
+  void RebalanceChildToRight(unsigned child_pos, unsigned count);
+
+  void MergeFromRight(Key_t key, BTreeNode* right);
+
+  void InsertItem(unsigned index, Key_t item) {
+    assert(index <= num_items_);
+    assert(index == 0 || Key(index - 1) < item);
+    assert(index == num_items_ || Key(index) > item);
+
+    ShiftRight(index);
+    SetKey(index, item);
+  }
+
+  uint8_t* KeyPtr(unsigned index) {
+    uint8_t* ptr = reinterpret_cast<uint8_t*>(this) + kKeyOffset + kKeySize * index;
+    return ptr;
+  }
+
+  const uint8_t* KeyPtr(unsigned index) const {
+    const uint8_t* ptr = reinterpret_cast<const uint8_t*>(this) + kKeyOffset + kKeySize * index;
+    return ptr;
+  }
+
+  union {
+    struct {
+      uint64_t num_items_ : 7;
+      uint64_t leaf_ : 1;
+      uint64_t : 56;
+    };
+
+    uint64_t val_;
+  };
+
+  static_assert(kMaxLeafKeys < 128, "num_items_ is 7 bits");
+};
+
+// Contains parent/index pairs. Meaning that node0->Child(index0) == node1.
+template <typename Policy> class BTreePath {
+  static constexpr unsigned kMaxDepth = 16;
+
+ public:
+  void Push(BTreeNode<Policy>* node, unsigned pos) {
+    assert(depth_ < kMaxDepth);
+    record_[depth_].node = node;
+    record_[depth_].pos = pos;
+    depth_++;
+  }
+
+  unsigned Depth() const {
+    return depth_;
+  }
+
+  std::pair<BTreeNode<Policy>*, unsigned> Last() const {
+    return {record_[depth_ - 1].node, record_[depth_ - 1].pos};
+  }
+
+  BTreeNode<Policy>* Node(unsigned i) const {
+    assert(i < depth_);
+    return record_[i].node;
+  }
+
+  unsigned Position(unsigned i) const {
+    assert(i < depth_);
+    return record_[i].pos;
+  }
+
+  void Pop() {
+    assert(depth_ > 0u);
+    depth_--;
+  }
+
+ private:
+  struct Record {
+    BTreeNode<Policy>* node;
+    unsigned pos;
+  };
+
+  Record record_[kMaxDepth];
+  unsigned depth_ = 0;
+};
+
+// Returns the position of the first item whose key is greater or equal than key.
+// if all items are smaller than key, returns num_items_.
+template <typename Policy> auto BTreeNode<Policy>::BSearch(Key_t key) const -> SearchResult {
+  uint32_t i = 0;
+  uint32_t n = num_items_;
+  typename Policy::KeyCompareTo cmp_op;
+  while (i < n) {
+    uint32_t j = (i + n) >> 1;
+    assert(j < n);
+
+    Key_t item = Key(j);
+
+    int cmp_res = cmp_op(key, item);
+    if (cmp_res == 0) {
+      return SearchResult{.index = j, .found = 1};
+    }
+
+    if (cmp_res < 0) {
+      n = j;
+    } else {
+      i = j + 1;  // we never return indices upto j because they are strictly less than key.
+    }
+  }
+  assert(i == n);
+
+  return {.index = i, .found = 0};
+}
```

Comment:
A little hard to read, instead of `i, n, j, >>1` I'd use `lo, hi, mid, /2` or smth else

But again, if we allow getting a Key_t* we can just use stl's lower_bound instead of re-writing it

Reply:
I will rename the variables. I do not want to use lower_bound because of inefficient `less` abstraction. I prefer a three-way compare because we gonna use it for sds strings.

Reply:
Hm, I see, so you want it to stop immediately if the item was found when placing the mid point

Reply:
I did not understand the question :)
I use `cmp_op` once per iteration, where lower_bound must use `compare` operator twice.